app/controllers/spree/affirm_controller.rb

```diff
@@ -0,0 +1,37 @@
+module Spree
+  class AffirmController < Spree::StoreController
+    protect_from_forgery except: [:confirm]
+
+    def confirm
+      order = current_order || raise(ActiveRecord::RecordNotFound)
+      checkout_token = affirm_params[:checkout_token]
+
+      if !checkout_token
+        return redirect_to checkout_state_path(order.state), notice: "Invalid order confirmation data passed in"
+      end
+
+      if order.complete?
+        return redirect_to spree.order_path(order), notice: "Order is already in complete state"
+      end
+
+      affirm_checkout = Solidus::Affirm::Checkout.new(token: checkout_token)
+
+      if affirm_checkout.valid?
+        affirm_checkout.save
+        order.payments.create!({
+          payment_method_id: affirm_params[:payment_method_id],
+          amount: order.total,
+          source: affirm_checkout
+        })
+        order.next
```

Comment:
As discussed, this is critical to the usability of the extension. We can't assert what a site intends to do after a payment has been authorized. And many sites don't use the state machine.

See https://stembolt.com/blog/class_eval-and-the-terrible-horrible-no-good-very-bad-day/ for an example on how to use composition here instead. We can define a default "handler" that the user can replace if necessary.